lib/preload.js

```diff
@@ -1,32 +1,196 @@
+    this.chunks.push(chunk)
+    next()
+  }
+
+  getData () {
+    return Buffer.concat(this.chunks)
+  }
+}
+
+const tarFiles = (dir, filenames) => {
+  let fullPath
+  const pack = tar.pack()
+  return Promise.map(filenames, (filePath) => {
+    fullPath = path.join(dir, filePath)
+    return Promise.join(fs.statAsync(fullPath), fs.readFileAsync(fullPath), (stats, data) => {
+      return pack.entryAsync({ name: filePath, size: stats.size, mode: stats.mode }, data)
+    })
+  })
+  .then(() => {
+    pack.finalize()
+    return pack
+  })
+}
+
+const bindMount = (source, target) => {
+  return {
+    Source: path.resolve(source),
+    Target: target,
+    Type: 'bind',
+    Consistency: 'delegated'
+  }
+}
+
+const createContainer = (docker, options) => {
+  const mounts = [bindMount(options.image, DISK_IMAGE_PATH_IN_DOCKER)]
+  if (options.splashImage) {
+    mounts.push(bindMount(options.splashImage, SPLASH_IMAGE_PATH_IN_DOCKER))
+  }
+  return docker.createContainer({
+    Image: DOCKER_IMAGE_TAG,
+    Name: preload.CONTAINER_NAME,
+    NetworkMode: 'host',
+    Privileged: true,
+    AttachStdout: true,
+    AttachStderr: true,
+    Env: [
+      `COMMAND=${options.command}`,
+      `APP_ID=${options.appId}`,
+      `API_TOKEN=${options.apiToken || ''}`,
+      `API_KEY=${options.apiKey || ''}`,
+      `COMMIT=${options.commit || ''}`,
+      `REGISTRY_HOST=${options.registryHost || ''}`,
+      `API_HOST=${options.apiHost || ''}`,
+      `DONT_DETECT_FLASHER_TYPE_IMAGES=${options.dontDetectFlasherTypeImages ? 'TRUE' : 'FALSE'}`,
+      `HTTP_PROXY=${options.proxy || ''}`,
+      `HTTPS_PROXY=${options.proxy || ''}`
+    ],
+    Mounts: mounts
+  })
+}
+
+const createContainerDisposer = (docker, options) => {
+  return createContainer(docker, options)
+  .disposer((container) => {
+    return container.remove()
+  })
+}
+
+const runContainer = (docker, container) => {
+  // We want to remove the container after it runs.
+  // We also want CTRL-C to kill the container and remove it.
+  const killContainer = (data) => {  // data is from stdin
+    if (data[0] !== CTRL_C) {
+      return
+    }
+    container.kill()
+  }
+  const removeListener = () => {
+    // Set things back to normal.
+    process.stdin.setRawMode(false)
+    process.stdin.pause()
+    process.stdin.removeListener('data', killContainer)
+  }
+  process.stdin.resume()
+  // Set raw mode to be able to intercept CTRL-C
+  process.stdin.setRawMode(true)
+  process.stdin.on('data', killContainer)
+
+  const stdout = new streamModule.PassThrough()
+  const stderr = new streamModule.PassThrough()
+  return container.start()
+  .then((container) => {
+    const statusCodePromise = new Promise((resolve, reject) => {
+      // Delete the container once it ends (or gets killed).
+      container.wait()
+      .then((info) => {
+        removeListener()
+        container.remove()
+        resolve(info.StatusCode)
+      })
+      .catch(reject)
+    })
```

Comment:
I'm pretty sure you can make this:
```javascript
const statusCodePromise = container.wait()
.then((info) => {
  removeListener()
  container.remove()
  return info.StatusCode
})
```

Also do we need to be waiting on `container.remove`?